src/www/app/outline_server.spec.ts

```diff
@@ -92,6 +92,216 @@ describe('OutlineServerRepository', () => {
       accessKey: shadowsocksConfigToAccessKey(CONFIG_1)
     });
   });
+
+  it('add stores servers', () => {
+    const storage = new InMemoryStorage();
+    const repo = new OutlineServerRepository(getFakeServerFactory(), new EventQueue(), storage);
+    const accessKey0 = shadowsocksConfigToAccessKey(CONFIG_0);
+    const accessKey1 = shadowsocksConfigToAccessKey(CONFIG_1);
+    repo.add(accessKey0);
+    repo.add(accessKey1);
+    const servers: ServersStorageV1 =
+        JSON.parse(storage.getItem(OutlineServerRepository.SERVERS_STORAGE_KEY));
+    expect(servers.length).toEqual(2);
+    expect(servers[0].accessKey).toEqual(accessKey0);
+    expect(servers[0].name).toEqual(CONFIG_0.name);
+    expect(servers[1].accessKey).toEqual(accessKey1);
+    expect(servers[1].name).toEqual(CONFIG_1.name);
+  });
+
+  it('add emits ServerAdded event', async () => {
+    const eventQueue = new EventQueue();
+    const repo =
+        new OutlineServerRepository(getFakeServerFactory(), eventQueue, new InMemoryStorage());
+    const accessKey = shadowsocksConfigToAccessKey(CONFIG_0);
+    repo.add(accessKey);
+    await new Promise<void>(resolve => {
+      eventQueue.subscribe(ServerAdded, (event: ServerAdded) => {
+        const server = event.server as OutlineServer;
+        expect(server.accessKey).toEqual(accessKey);
+        expect(server.name).toEqual(CONFIG_0.name);
+        resolve();
+      });
+      eventQueue.startPublishing();
+    });
+  });
+
+  it('add throws on invalid access keys', () => {
+    const repo = new OutlineServerRepository(
+        getFakeServerFactory(), new EventQueue(), new InMemoryStorage());
+    expect(() => repo.add('ss://invalid')).toThrow();
```

Comment:
Do we have concrete error types that we can check for?

Reply:
Done in b67381b. I hhad tried `toThrow`, which compares error instances; now using `toThrowError`, which takes an error constructor.